### PR TITLE
[CI-macos]: add caching for ghcup & cabal (reduce build time by 4 minutes)

### PR DESCRIPTION
**Pull request by ak3n**
based on #1920

### CODE REVIEW

File: .github/workflows/macos.yaml [hunk 1]
```diff
@@ -16,33 +16,58 @@ jobs:
       matrix:
         ghc: ["9.6.4"]
         cabal: ["3.10"]
-        os: ["macOS-latest"]
```
Comment (emilypi) [inline .github/workflows/macos.yaml:19]: we should do both - one goes through the macos github runner, but one is fixed hardware we have onhand for M1 specifically for regression.

**Review by emilypi — COMMENTED** [review]
LGTM other than the one comment

**Review by emilypi — APPROVED** [review]

File: .github/workflows/macos.yaml (hunk 1 above)
Comment (emilypi) [reply .github/workflows/macos.yaml:19]: Checked and Macos-latest is macos-14

**Comment by edmundnoble** [comment]
Would this .ghcup cache also work fine for our other builds?

**Pull request merged**